app/src/main/java/com/cakii/nimble_android_challenge/data/interceptor/AuthInterceptor.kt

```diff
@@ -0,0 +1,46 @@
+package com.cakii.nimble_android_challenge.data.interceptor
+
+import android.content.Context
+import com.cakii.nimble_android_challenge.R
+import com.cakii.nimble_android_challenge.data.entity.Auth
+import com.cakii.nimble_android_challenge.data.entity.AuthRequest
+import com.cakii.nimble_android_challenge.utils.Prefs
+import com.google.gson.Gson
+import okhttp3.*
+import java.io.IOException
+import java.net.URL
+
+
+class AuthInterceptor(private val context: Context, private val prefs: Prefs): Authenticator {
+
+    @Throws(IOException::class)
+    override fun authenticate(route: Route?, response: Response?): Request? {
+        return if (response?.code() == 401) {
+
+            val client = OkHttpClient()
+            val url = URL("${context.getString(R.string.base_url)}/oauth/token")
+
+            val authRequest = AuthRequest("carlos@nimbl3.com", "antikera", "password")
+            val requestJson = Gson().toJson(authRequest)
```

Comment:
the idea of oauth2, is not to keep your credential (username + password) anywhere after the login, but instead, to reauthenticate yourself, you should use the refreshToken we provided 😄 that's the pattern we're looking at
https://medium.com/@sandeeptengale/problem-solved-2-access-token-refresh-with-okhttp-authenticator-5ccb798ede70

Reply:
actually, I just revise the requirement and we don't provide refresh token for now to simplify the login flow. So instead, the recommendation should be not to hard coding the username/password here, but instead, we would rather inject the component where you can get the credential from 😄